File: ios/runanywhere-sdks-latest/sdk/runanywhere-web/packages/core/src/Public/Extensions/Hybrid/HybridDeviceState.ts
/**
 * HybridDeviceState.ts
 *
 * Web binding for the cross-SDK host device-state vtable
 * (rac_hybrid_device_state.h) and the named custom-filter callback table
 * (rac_hybrid_custom_filter.h). Mirrors Swift's `HybridDeviceState.swift` +
 * `HybridCustomFilter.swift` and Kotlin's `DeviceStateProvider` +
 * `CustomFilterPredicate` wiring.
 *
 * All routing LOGIC stays in commons — these helpers only install host
 * callbacks (is_online / battery_percent / is_thermal_throttled) and register
 * named predicates so commons can resolve + invoke them during candidate
 * filtering. The host never pre-filters or toggles router slots.
 *
 * Callbacks cross the WASM boundary as function-table indices
 * (Module.addFunction), exactly like the download/HTTP progress callbacks. The
 * binding targets the flattened `*_from_js` commons wrappers (see
 * HybridWasmModule.ts) so JS never has to know the C struct / ctx layout. When
 * those exports are absent (current Web WASM builds), registration is a
 * graceful no-op returning false — the router still works against the commons
 * optimistic default (always-online, 100%, not-throttled).
 */

import { SDKLogger } from '../../../Foundation/SDKLogger.js';
import { RAC_OK } from '../../../Foundation/RACErrors.js';
import {
  getModuleForCapability,
} from '../../../runtime/EmscriptenModule.js';
import type { HybridWasmModule } from './HybridWasmModule.js';

const logger = new SDKLogger('Hybrid.DeviceState');

/**
 * Host-supplied source of the device state the hybrid router needs.
 * Implementations MUST be reentrant: commons may invoke these from request
 * threads. Mirrors Swift's `HybridDeviceStateProvider` /
 * Kotlin's `DeviceStateProvider`.
 */
export interface HybridDeviceStateProvider {
  /** True iff the host has a usable internet connection right now. */
  isOnline(): boolean;
  /** Battery level in [0, 100]; return 100 on hosts without a battery. */
  batteryPercent(): number;
  /** True when the device is currently thermally throttled. */
  isThermalThrottled(): boolean;
}

/** Default browser-shaped provider built from `navigator` facts. `navigator`
 * exposes no battery synchronously, so battery is reported as 100 and thermal
 * as false — matching the commons optimistic default but tracking real
 * online/offline transitions via `navigator.onLine`. */
export function browserDeviceStateProvider(): HybridDeviceStateProvider {
  return {
    isOnline: () =>
      typeof navigator !== 'undefined' && typeof navigator.onLine === 'boolean'
        ? navigator.onLine
        : true,
    batteryPercent: () => 100,
    isThermalThrottled: () => false,
  };
}

function hybridModule(): HybridWasmModule | null {
  return (getModuleForCapability('stt') ??
    getModuleForCapability('commons')) as HybridWasmModule | null;
}

// ── Device-state vtable installer ──────────────────────────────────────────

let installedDeviceStatePtrs: number[] = [];

/** Register `provider` as the host device-state source the router consults on
 * every transcribe. Pass `null` to unregister and fall back to the commons
 * optimistic default. Returns true when the native vtable was installed. */
export function setHybridDeviceStateProvider(
  provider: HybridDeviceStateProvider | null,
): boolean {
  const mod = hybridModule();
  if (!mod || typeof mod.addFunction !== 'function' || typeof mod.removeFunction !== 'function') {
    logger.warning('module missing addFunction/removeFunction; cannot install device-state vtable');
    return false;
  }
  const setFn = mod._rac_hybrid_set_device_state_from_js;
  if (typeof setFn !== 'function') {
    logger.warning(
      'WASM module does not export _rac_hybrid_set_device_state_from_js; ' +
        'device-state filters use the commons optimistic default. See ' +
        'HybridWasmModule.ts BUILD DELTA.',
    );
    return false;
  }

  // Retire any previously installed callbacks AFTER swapping the vtable so no
  // request thread is mid-callback against a freed table slot.
  const previousPtrs = installedDeviceStatePtrs;
  installedDeviceStatePtrs = [];

  if (!provider) {
    setFn(0, 0, 0); // restore default
    for (const ptr of previousPtrs) mod.removeFunction(ptr);
    return true;
  }

  // is_online → 'ip' (returns rac_bool_t/i32, takes user_data pointer);
  // battery_percent → 'ip' (returns i32); is_thermal_throttled → 'ip'.
  const isOnlinePtr = mod.addFunction(() => (provider.isOnline() ? 1 : 0), 'ip');
  const batteryPtr = mod.addFunction(() => provider.batteryPercent() | 0, 'ip');
  const thermalPtr = mod.addFunction(() => (provider.isThermalThrottled() ? 1 : 0), 'ip');

  const rc = setFn(isOnlinePtr, batteryPtr, thermalPtr);
  if (rc !== RAC_OK) {
    mod.removeFunction(isOnlinePtr);
    mod.removeFunction(batteryPtr);
    mod.removeFunction(thermalPtr);
    logger.error(`rac_hybrid_set_device_state_from_js failed: rc=${rc}`);
    return false;
  }
  installedDeviceStatePtrs = [isOnlinePtr, batteryPtr, thermalPtr];
  for (const ptr of previousPtrs) mod.removeFunction(ptr);
  return true;
}

// ── Custom-filter table ────────────────────────────────────────────────────

/** name → function-table index, so each predicate's table slot can be retired
 * on unregister. */
const customFilterPtrs = new Map<string, number>();

/** Register (or replace) a named custom-filter predicate with commons. The
 * predicate is invoked by commons (NOT JS) once per candidate during filtering;
 * it receives the candidate model id. Returns true on success. */
export function registerHybridCustomFilter(
  name: string,
  check: (modelId: string) => boolean,
): boolean {
  if (!name) {
    logger.error('custom-filter name must be non-empty');
    return false;
  }
  const mod = hybridModule();
  if (!mod || typeof mod.addFunction !== 'function' || typeof mod.removeFunction !== 'function') {
    return false;
  }
  const registerFn = mod._rac_hybrid_register_custom_filter_from_js;
  if (typeof registerFn !== 'function' || typeof mod.UTF8ToString !== 'function') {
    logger.warning(
      'WASM module does not export _rac_hybrid_register_custom_filter_from_js; ' +
        'custom filters are ignored. See HybridWasmModule.ts BUILD DELTA.',
    );
    return false;
  }

  // Trampoline: commons passes the candidate model id as a C-string (the
  // flattened from_js wrapper extracts ctx.candidate_model_id), we dispatch to
  // the JS predicate and map the boolean to rac_bool_t. Signature 'ip': returns
  // rac_bool_t/i32, takes the model-id char* pointer.
  const predicatePtr = mod.addFunction((modelIdPtr: number) => {
    const modelId = modelIdPtr ? mod.UTF8ToString!(modelIdPtr) : '';
    return check(modelId) ? 1 : 0;
  }, 'ip');

  const namePtr = allocCString(mod, name);
  try {
    const rc = registerFn(namePtr, predicatePtr);
    if (rc !== RAC_OK) {
      mod.removeFunction(predicatePtr);
      logger.error(`rac_hybrid_register_custom_filter_from_js('${name}') failed: rc=${rc}`);
      return false;
    }
  } finally {
    mod._free?.(namePtr);
  }

  // Replace any prior predicate registered under the same name + retire its slot.
  const previous = customFilterPtrs.get(name);
  if (previous !== undefined) mod.removeFunction(previous);
  customFilterPtrs.set(name, predicatePtr);
  return true;
}

/** Remove a named custom-filter predicate. No-op when not registered. */
export function unregisterHybridCustomFilter(name: string): boolean {
  if (!name) return false;
  const mod = hybridModule();
  if (!mod) return false;
  const unregisterFn = mod._rac_hybrid_unregister_custom_filter;
  if (typeof unregisterFn === 'function') {
    const namePtr = allocCString(mod, name);
    try {
      unregisterFn(namePtr);
    } finally {
      mod._free?.(namePtr);
    }
  }
  const ptr = customFilterPtrs.get(name);
  if (ptr !== undefined && typeof mod.removeFunction === 'function') {
    mod.removeFunction(ptr);
  }
  customFilterPtrs.delete(name);
  return true;
}

function allocCString(mod: HybridWasmModule, value: string): number {
  const len = mod.lengthBytesUTF8!(value) + 1;
  const ptr = mod._malloc!(len);
  mod.stringToUTF8!(value, ptr, len);
  return ptr;
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/hybrid/HybridDeviceState.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Kotlin binding for the cross-SDK host device-state vtable
 * (rac_hybrid_device_state.h). The hybrid router consults this vtable on
 * every transcribe() to evaluate the Network / Battery hard filters; it is
 * NOT passed in the per-request routing context.
 *
 * All routing LOGIC stays in commons — this file only installs the host
 * callbacks (isOnline / batteryPercent / isThermalThrottled) through the JNI
 * thunk `racHybridSetDeviceState`, which wires the Kotlin provider into the
 * `rac_hybrid_device_state_ops_t` vtable in commons.
 *
 * Mirrors Swift's HybridDeviceState.swift (HybridDeviceStateProvider protocol
 * + HybridDeviceState.setProvider/clear).
 */

package com.runanywhere.sdk.hybrid

import com.runanywhere.sdk.infrastructure.logging.SDKLogger
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge

/**
 * Host-supplied source of the device state the hybrid router needs.
 *
 * Implementations MUST be thread-safe / reentrant: commons may invoke these
 * from multiple request threads concurrently (see the warning in
 * rac_hybrid_device_state.h).
 *
 * Method names MUST match what the JNI thunk looks up via `GetMethodID`
 * (`isOnline()Z`, `batteryPercent()I`, `isThermalThrottled()Z`) in
 * sdk/runanywhere-commons/src/jni/rac_hybrid_device_state_jni.cpp — do NOT
 * rename without also updating the C++ side.
 */
interface HybridDeviceStateProvider {
    /** True iff the host has a usable internet connection right now. */
    fun isOnline(): Boolean

    /** Battery level in `[0, 100]`; return 100 on hosts without a battery. */
    fun batteryPercent(): Int

    /** True when the device is currently thermally throttled. */
    fun isThermalThrottled(): Boolean
}

/**
 * Installs / clears the device-state vtable in commons.
 *
 * Exactly one provider is active process-wide (the C ABI holds a single
 * vtable). Re-registering replaces the previous provider. Mirrors Swift's
 * `HybridDeviceState` enum.
 */
object HybridDeviceState {
    private val logger = SDKLogger("Hybrid.DeviceState")

    /**
     * Register [provider] as the host device-state source the router consults
     * on every transcribe. Pass `null` to unregister and fall back to the
     * commons optimistic default (always-online, 100% battery, not-throttled).
     *
     * Typical Android wiring (call once after SDK init):
     *
     *     HybridDeviceState.setProvider(AndroidDeviceStateProvider(applicationContext))
     *
     * Thread-safe; the native vtable swap is atomic.
     *
     * @return `true` when the vtable was installed (or cleared) successfully.
     */
    @JvmStatic
    fun setProvider(provider: HybridDeviceStateProvider?): Boolean {
        if (provider == null) {
            return clear()
        }
        RunAnywhereBridge.ensureNativeLibraryLoaded()
        val rc = RunAnywhereBridge.racHybridSetDeviceState(provider)
        if (rc != RunAnywhereBridge.RAC_SUCCESS) {
            logger.error("racHybridSetDeviceState failed: rc=$rc")
            return false
        }
        return true
    }

    /**
     * Detach the host provider and restore the commons default vtable.
     *
     * @return `true` when the default vtable was restored successfully.
     */
    @JvmStatic
    fun clear(): Boolean {
        RunAnywhereBridge.ensureNativeLibraryLoaded()
        val rc = RunAnywhereBridge.racHybridSetDeviceState(null)
        if (rc != RunAnywhereBridge.RAC_SUCCESS) {
            logger.error("racHybridSetDeviceState(null) failed: rc=$rc")
            return false
        }
        return true
    }
}
